### PR TITLE
[SDESK-7275] Support versioned async resources

### PR DESCRIPTION
**Pull request by MarkLark86**

### CODE REVIEW

**Review by eos87 — APPROVED** [review]
It looks good to me, though I feel I may not have the full context or understanding needed for a thorough review. It might be helpful to have another set of eyes on this. @petrjasek, could you take a quick look? Thanks!

File: superdesk/core/mongo.py [hunk 1]
```diff
@@ -240,104 +246,138 @@ def stop(self):
...
+        indexes = mongo_config.indexes or []
+        if indexes:
+            self.create_collection_indexes(
+                self.get_collection(resource_name, versioning=False),
+                indexes,
+                ignore_duplicate_keys,
+            )
+
+        if mongo_config.versioning:
+            indexes += (mongo_config.version_indexes or []) + [
```
Comment (petrjasek) [inline superdesk/core/mongo.py:321]: I'm not sure if we need to add any indexes from the main resource here, imo that `_id_document` should do in theory

File: superdesk/resource_fields.py [hunk 2]
```diff
@@ -22,5 +23,6 @@
 ITEMS = "_items"
 LINKS = "_links"
 ETAG = "_etag"
-VERSION = "_current_version"
+VERSION = CURRENT_VERSION = "_current_version"
+LATEST_VERSION = "_latest_version"
```
Comment (petrjasek) [inline superdesk/resource_fields.py:27]: I wonder what's the usecase for this field, seems like something that would require an update to all previous versions on each save, but at the end it's just duplicating the `_current_version` value from the main collection

**Review by petrjasek — COMMENTED** [review]
would it register that extra `{resource}_version` resource as it does currently? I see it in a way useful that you can use it as any other resource, but it might not be actually what we should allow, maybe we can have some specific base class for those which would be readonly via the api

**Comment by MarkLark86** [comment]
> would it register that extra `{resource}_version` resource as it does currently?

No it doesn't. Instead it provides an `version: int | None` argument to the [find_one](https://github.com/superdesk/superdesk-core/pull/2678/files#diff-7afef0e188cec0364bc5d9e62317fe6639b375a9889f923817c3ba2aed3777feR118) method on the AsyncResourceService, along with a `get_all_item_versions` and `get-item_version` methods. It is also handled by requests through the REST API via the same `version` URL argument we use now.

> maybe we can have some specific base class for those which would be readonly via the api

The AsyncResourceService doesn't publicly expose any interface to create, modify nor delete a version document. It is internally handled in the create, update delete methods on that service.

File: superdesk/core/mongo.py (hunk 1 above)
Comment (MarkLark86) [reply superdesk/core/mongo.py:321]: I just copied what Eve did. We should at a minimum have indexes for `_id_document_` and `_current_version`. I'll make this change tomorrow

File: superdesk/resource_fields.py (hunk 2 above)
Comment (MarkLark86) [reply superdesk/resource_fields.py:27]: It's populated on fetch (so no need to update previous items), and is also used in the front-end.

Comment (petrjasek) [reply superdesk/resource_fields.py:27]: when fetching previous versions?

File: superdesk/core/mongo.py (hunk 1 above)
Comment (petrjasek) [reply superdesk/core/mongo.py:321]: yep, we need `_id_document` for sure, I'm just not sure about the other ones

File: superdesk/resource_fields.py (hunk 2 above)
Comment (MarkLark86) [reply superdesk/resource_fields.py:27]: Correct

File: superdesk/core/mongo.py [hunk 3]
```diff
@@ -240,104 +247,138 @@ def stop(self):
...
+
+        if mongo_config.versioning:
+            indexes = (mongo_config.version_indexes or []) + [
+                MongoIndexOptions(
+                    name="_id_document_1",
+                    keys=[(VERSION_ID_FIELD, 1)],
+                    background=True,
+                    unique=False,
+                ),
+                MongoIndexOptions(
```
Comment (petrjasek) [inline superdesk/core/mongo.py:329]: imo this one might be enough, that should be usable also for lookups using just `_id_document`

**Review by petrjasek — APPROVED** [review]

**Pull request merged**